SymIntify cat and narrow

wconstab opened the pull request.

wconstab committed:
SymIntify cat and narrow

ezyang commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -1098,7 +1098,7 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
   // NB: static_cast because there's another variant of narrow. However, we don't
   // want to support the other variant yet bc it isn't documented...
-  m.impl("narrow", static_cast<Tensor(*)(const Tensor&,int64_t,int64_t,int64_t)>(native::narrow)); // composite wrt autograd
+  m.impl("narrow", static_cast<Tensor(*)(const Tensor&,int64_t,SymInt,SymInt)>(native::narrow_symint)); // composite wrt autograd
```
Feels like the cast may not be necessary anymore

wconstab replied:
i think you're right.

ezyang commented on aten/src/ATen/native/transformers/cuda/attention.cu:
```diff
@@ -387,7 +387,7 @@ __host__ std::tuple<Tensor, Tensor, Tensor> transform_bias_rescale_qkv_cuda(
           auto offsets =
               NestedTensor_batch_offsets_from_size_tensor(sizes, sizes.numel());
-          at::native::narrow(offsets, 0, sizes.numel() + 1, sizes.numel())
+          at::native::narrow_symint(offsets, 0, sizes.numel() + 1, sizes.numel())
```
You may want to do the same thing we did with reshape() and keep a non symint native::narrow around so you don't have to go update all these call sites

wconstab replied:
is the non symint native::narrow just declared manually in a non codegen header or is there a trick (i couldn't find anything looking at native_functions.yaml for reshape)

wconstab replied:
ok i tried this but i got undefined narrow in libtorch_cuda.so, when I included BC header in `aten/src/ATen/native/transformers/cuda/attention.cu`. I guess there is probably something else I need to change about the build to make this work?

wconstab replied:
i gave up on this part bc its just a couple callsites. could revisit if needed.

ezyang replied:
Did you put TORCH_API on the native:: header, and properly include it at the definition site?

ezyang commented on c10/core/WrapDimMinimal.h:
```diff
@@ -22,4 +25,18 @@ static inline int64_t maybe_wrap_dim(
+  // Inline the fast paths
+  if (C10_LIKELY(dim_post_expr * -1 <= dim && dim < dim_post_expr)) {
+    // Branch-less version of dim + (dim < 0 ? dim_post_expr : 0)
+    return dim + dim_post_expr * (dim < 0);
```
@Chillee can you take a look at these

Chillee replied:
This doesn't matter, since these are on dims (which are fixed anyways).

wconstab replied:
eh.. check the call site of maybe_wrap_dim inside narrow.  and check the call site of narrow from cat.

Chillee replied:
Isn't this a hack lol? This was written for dim wrapping, but the call-site is just using it for shape wrapping.

In general, I'd probably prefer just having a guard here. Technically, that's what this does today (since boolean comparisons immediately guard), but I'd be fine with writing this in the more natural way.

wconstab replied:
i suspect you're right. I probably should have left the dim arg as int, and put a guard in cat.  (is that what you meant?)

wconstab replied:
although tbc, would I still want dim_post_expr to be sym, and keep (ab)using the wrap_dim for wrap_shape, or not?

Chillee replied:
> although tbc, would I still want dim_post_expr to be sym, and keep (ab)using the wrap_dim for wrap_shape, or not?

Yeah I think that's fine. Basically, TL;DR: Rewrite this to more explicitly use control flow.

ezyang commented on c10/core/WrapDimMinimal.h:
```diff
@@ -22,4 +25,18 @@ static inline int64_t maybe_wrap_dim(
   return c10::detail::maybe_wrap_dim_slow(dim, dim_post_expr, wrap_scalar);
 }
 
+static inline SymInt maybe_wrap_dim_symint(
```
Consider templating the original to avoid code duplication

ezyang commented on c10/core/WrapDimMinimal.cpp:
```diff
@@ -35,5 +35,38 @@ int64_t maybe_wrap_dim_slow(
       false, "should never reach here as dim should be out-of-bounds");
 }
 
+SymInt maybe_wrap_dim_slow_symint(
```
Consider templating

ezyang commented on aten/src/ATen/WrapDimUtils.h:
```diff
@@ -97,6 +108,18 @@ static inline int64_t legacy_cat_wrap_dim(
+    }
+    return maybe_wrap_dim(dim, sizes.size());
+  }
+  return dim;
```
Consider templating

wconstab committed:
Sort Tensor args before SymInt/int args for argparser

- ensure that when 2 schemas/overloads are otherwise
  the same except one has a Tensor and the other has a
  (Symint or int), we pick the Tensor one.
- This is important becuase Tensors of numel 1 can be converted
  to SymInt/int using __index__, but ops with explicit Tensor
  arg inputs should get priority so they can check dim or other
  properties before implicit conversion happens

SymIntify narrow.Tensor

- also restore non-symint narrow functions for BC

Templatify maybe_wrap_dim_slow

wconstab commented on aten/src/ATen/functorch/BatchRulesDecompositions.cpp:
```diff
@@ -168,7 +168,7 @@ TORCH_LIBRARY_IMPL(aten, FuncTorchBatched, m) {
   OP_DECOMPOSE(msort);
   OP_DECOMPOSE(mT);
-  OP_DECOMPOSE(narrow);
+  m.impl("narrow", native::narrow_symint);
```
@ezyang @Chillee i'm not sure what this actually does.  It compiles but i'm not happy with the fact that it only handles one of the 2 overloads, and i didn't figure out the right incantation to try to handle both of them.

Chillee replied:
I don't think it matters too much. Basically, we're just registering vmap to decompose compositeimplicitautograd ops.

ezyang replied:
this is fine, I think only hitting the non-Tensor overload is fine, esp if no ci fails. I think technically probably the tensor overload should get a decompose too but I'll file this as preexisting condition

ezyang commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -1096,9 +1096,7 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
   m.impl("movedim.intlist", movedim_batching_rule);
   m.impl("movedim.int", static_cast<Tensor(*)(const Tensor&,int64_t,int64_t)>(native::movedim)); // composite wrt autograd
-  // NB: static_cast because there's another variant of narrow. However, we don't
-  // want to support the other variant yet bc it isn't documented...
```
Half of this comment is still relevant, btw

ezyang commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1125,22 +1125,37 @@ Tensor& narrow_copy_dense_cpu_out(
+    Until we support passing a SymInt through a singleton tensor, we're stuck fully-specializing
+    this overload.
+  */
+  st.guard_int(__FILE__, __LINE__);
```
this guard doesn't do anything lol

ezyang commented on c10/core/WrapDimMinimal.cpp:
```diff
@@ -3,10 +3,17 @@
+    bool wrap_scalar);
+
+template <typename T>
+T maybe_wrap_dim_slow(T dim, T dim_post_expr, bool wrap_scalar) {
```
I want to point out why sometimes you were able to template, and why sometimes you weren't. C++ will attempt implicit conversions when calling functions. However, if the function is templated, no implicit conversions will happen. It turns out that maybe_wrap_dim_slow has no implicit conversions, which is why this is ok

wconstab replied:
yea, i realized that we have a bunch of implicit conversions from .sizes() calls coming into this function, so i gave up for now

ezyang commented on c10/core/WrapDimMinimal.h:
```diff
@@ -1,25 +1,45 @@
-maybe_wrap_dim_slow(int64_t dim, int64_t dim_post_expr, bool wrap_scalar);
-}
+template <typename T>
+C10_API T maybe_wrap_dim_slow(T dim, T dim_post_expr, bool wrap_scalar);
```
declaring a template but not giving any implementations of it doesn't do anything

ezyang replied:
oh you're playing tricks with defining the template instantiations in the cpp file lol

ezyang replied:
idk if this is idiomatic or not, but I certainly learned a new C++ feature today!

ezyang commented on c10/core/WrapDimMinimal.cpp:
```diff
@@ -3,10 +3,17 @@
+template SymInt maybe_wrap_dim_slow(
+    SymInt dim,
+    SymInt dim_post_expr,
+    bool wrap_scalar);
```
Hmm, I think you need to annotate these with `C10_API`, and that's why you got linker errors when you tried to template

ezyang committed:
CR fixes

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
@pytorchbot merge

ezyang committed:
lintfix

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
@pytorchbot merge

wconstab commented on c10/core/WrapDimMinimal.cpp:
```diff
@@ -42,5 +33,15 @@ T maybe_wrap_dim_slow(T dim, T dim_post_expr, bool wrap_scalar) {
       false, "should never reach here as dim should be out-of-bounds");
 }
 
+// Explicitly instantiate the template at the two types it will be used
```
I could guess why C10_API is important for linker errors across lib boundaries,

is moving it below also important for some reason or just conventional?

ezyang replied:
nah, it's just cuz I feel it's weird to instantiate a template before you actualy defined it

ezyang committed:
Include header in cuda

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
@pytorchbot merge

wconstab committed:
Include one more missing header

wconstab commented:
@pytorchbot merge

ezyang committed:
Update TensorShape.cpp

wconstab committed:
Merge branch 'master' into wconstab/master_catnarrow

ezyang committed:
xfail update

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
@pytorchbot merge -f "previous ci check was ok except for xfail update"

Flamefire commented:
This has broken master due to missing import of `SymIntT` (used e.g. in line 831
